xml_converter/src/xml_converter.cpp

```diff
@@ -414,45 +408,46 @@ bool filename_comp(string a, string b) {
     return lowercase(a) < lowercase(b);
 }
 
-vector<string> get_xml_files(string directory) {
+vector<string> get_xml_files(string directory, string data_directory) {
     vector<string> files;
-    vector<string> subfolders;
-    string path;
     DIR* dir = opendir(directory.c_str());
-    struct dirent* entry = readdir(dir);
+    struct dirent* entry;
     while ((entry = readdir(dir)) != NULL) {
-        if (entry->d_type == DT_DIR && has_suffix(directory, ".") == false) {
-            path = directory + "/";
-            path += entry->d_name;
-            subfolders.push_back(path);
-        }
-        else if (has_suffix(entry->d_name, ".xml")) {
-            path = directory + "/";
-            path += entry->d_name;
-            files.push_back(path);
-        }
-        else {
-            continue;
+        string filename = entry->d_name;
+        if (filename != "." && filename != "..") {
+            string path = directory + "/" + filename;
+            if (entry->d_type == DT_DIR) {
+                string new_folder = data_directory + "/" + filename;
+                if (mkdir(new_folder.c_str(), 0700) == -1 && errno != EEXIST) {
+                    cout << "Error making " << new_folder << endl;
+                    continue;
+                }
+                vector<string> subfiles = get_xml_files(path, new_folder);
+                for (string subfile : subfiles) {
+                    cout << subfile << endl;
+                    files.push_back(subfile);
+                }
+            }
+            else if (entry->d_type == DT_REG) {
+                if (has_suffix(filename, ".xml")) {
+                    files.push_back(path);
+                }
+                else if (has_suffix(filename, ".trl") == false) {
+                    string new_path = data_directory + "/" + filename;
+                    ifstream infile(path, ios::binary);
+                    ofstream outfile(new_path, ios::binary);
+                    outfile << infile.rdbuf();
+                }
+            }
         }
     }
     closedir(dir);
-
     std::sort(files.begin(), files.end(), filename_comp);
-    std::sort(subfolders.begin(), subfolders.end(), filename_comp);
-
-    for (const std::string& subfolder : subfolders) {
-        if (has_suffix(subfolder, ".") == false && has_suffix(subfolder, "..") == false) {
-            vector<string> subfiles = get_xml_files(subfolder);
-            files.insert(files.end(), subfiles.begin(), subfiles.end());
-        }
-    }
     return files;
 }
 
-void convert_taco_directory(string directory, map<string, Category>* marker_categories, vector<Parseable*>* parsed_pois) {
-    string data = "rsync -a " + directory + "/Data ./protobins";
-    system(data.c_str());
-    vector<string> xml_files = get_xml_files(directory);
+void convert_taco_directory(string directory, string data_directory, map<string, Category>* marker_categories, vector<Parseable*>* parsed_pois) {
```

Comment:
data_directory as above

Make all the output stuff a different function below

Reply:
This still requires an `output_directory` meaning that it is still not possible to read the xml without also moving the data as you have included a call to `move_supplimentary_files()` inside of `convert_taco_directory()`. Maybe convert_taco_directory is now misnamed as it used to just parse and convert into in-memory datums.

Remember we need to be able to

* Read XML (and lint it)
* Write XML
* Read Protobuf
* Write Protobuf
* Write Nothing

in any combination of read and writes

* **Read XML -> Write XML** Lint and auto sanitize the xml files
* **Read XML -> Write Protobuf** Convert the xml to proto for burrito
* **Read XML -> Write Nothing** Lint but dont auto anything
* **Read Protobuf -> Write XML** Convert proto to xml to share
* **Read Protobuf -> Write Protobuf** It is binary data humans cant read but this step would still theoretically sanitize it
* **Read Protobuf -> Write Nothing** We dont lint the proto so this might be useless

Additionally we will eventually have 2 different proto "formats". They would use the same proto message but one would be split by mapid and the other would be a big ol' ball of everything which could be used to share to other users just like how `.taco` files currently work. So making sure the system is flexible enough to be able to be extended without having to edit hard to discover locations is good practice.